### ISSUE TITLE
Doxygen will not generate class information for a C# class that contains certain string combinations

### ISSUE THREAD

**Issue by gvsoft3750** [doxygen/doxygen#9736]
Doxygen fails to generate any class information for a class if the source file contains certain string combinations:

```
namespace MyNamespace
{
    public class MyClass
    {
        public string MyMethod()
        {
            return $"'{"''"}'"; // fails (class information is NOT generated)
        }
    }
}
```

When no class information is generated:

<img width="270" alt="NoClassInfoGenerated" src="https://user-images.githubusercontent.com/98843223/209022100-81a52401-086d-4c95-91dc-2a32db412479.PNG">

Here is what is expected:

<img width="297" alt="ClassInfoGenerated" src="https://user-images.githubusercontent.com/98843223/209022157-d91cae6f-8c77-4b84-a357-b2cb0a418788.PNG">

Replacing the above return statement with the following gives the results as described in the comment:

`return $"'{"''"}'"; // fails (class information is NOT generated)`
`return $"'{"'"}'";  // fails (class information is NOT generated)`
`return $"'{""}'";   // okay (class information is generated)`
`return $"'{"''"}";  // okay (class information is generated)`
`return $"'{""}'";   // okay (class information is generated)`

The original code where I discovered this issue looked like this:

`returnValue = $"'{value.Replace("'", "''")}'";`

I am using Doxygen v1.9.5 (obtained via download), on Windows 10, 64-bit

I was able to work around the problem in my original code by doing this:

```
string replaced = value.Replace("'", "''");
returnValue = $"'{replaced}'";
```

Dennis Jones

**Comment by albert-github** [comment]
How would the resulting return value look like?

**Comment by gvsoft3750** [comment]
The statement is intended to take a C-style string and convert it into a form that can be stored in SQL.  That is, it replaces any embedded single-quotes by doubling them, and then wraps the entire string in single-quotes.  For example, a possible result might be: 'Table ''Item''  is empty'.Hope that is clear.DennisSent from my T-Mobile 4G LTE Device
-------- Original message --------From: albert-github ***@***.***> Date: 12/22/22  1:34 AM  (GMT-08:00) To: doxygen/doxygen ***@***.***> Cc: gvsoft3750 ***@***.***>, Author ***@***.***> Subject: Re: [doxygen/doxygen] Doxygen will not generate class information for a C# class that contains certain string combinations (Issue #9736) 
How would the resulting return value look like?

**Comment by albert-github** [comment]
It looks like we have to do here with so called interpolated strings in C# which have a bit of a (in C-terms) peculiar form as when we look at:
```
$"'{"''"}'"
``` 
As far as I understand it running through the string:
- the `$"` starts the interpolated string
- the `'` is a regular character 
- the `{` starts the part to be "interpolated"
- the `"''"` is a string in to be "interpolated"
- the `}` ends the part to be "interpolated"
- the `'` is a regular character 
- the `"` ends the interpolated string

Is this a good summary?

**Comment by gvsoft3750** [comment]
You are correct. This is C#, and it is an example of string interpolation.  Your summary is reasonable.  This is a contrived example, however, as there would never be any reason to put two single-quote characters inside the interpolated section (within the curly braces).  I did include the real code in my original post.DennisSent from my T-Mobile 4G LTE Device
-------- Original message --------From: albert-github ***@***.***> Date: 12/22/22  3:42 AM  (GMT-08:00) To: doxygen/doxygen ***@***.***> Cc: gvsoft3750 ***@***.***>, Author ***@***.***> Subject: Re: [doxygen/doxygen] Doxygen will not generate class information for a C# class that contains certain string combinations (Issue #9736) 
It looks like we have to do here with so called interpolated strings in C# which have a bit of a (in C-terms) peculiar form as when we look at:
$"'{"''"}'"

As far as I understand it running through the string:

the $" starts the interpolated string
the ' is a regular character
the { starts the part to be "interpolated"
the "''" is a string in to be "interpolated"
the } ends the part to be "interpolated"
the ' is a regular character
the " ends the interpolated string

Is this a good summary?

**Comment by albert-github** [comment]
Looks like after numerous years a new version of the C# Language Definition (Standard ECMA-334 6th Edition i.e. https://www.ecma-international.org/wp-content/uploads/ECMA-334_6th_edition_june_2022.pdf) has been released.

**Comment by gvsoft3750** [comment]
Interpolated strings are not new to C#.  They've been in the language for several years.DennisSent from my T-Mobile 4G LTE Device
-------- Original message --------From: albert-github ***@***.***> Date: 12/22/22  3:59 AM  (GMT-08:00) To: doxygen/doxygen ***@***.***> Cc: gvsoft3750 ***@***.***>, Author ***@***.***> Subject: Re: [doxygen/doxygen] Doxygen will not generate class information for a C# class that contains certain string combinations (Issue #9736) 
Looks like after numerous years a new version of the C# Language Definition (Standard ECMA-334 6th Edition i.e. https://www.ecma-international.org/wp-content/uploads/ECMA-334_6th_edition_june_2022.pdf) has been released.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9738

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.5.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).

**Comment by gvsoft3750** [comment]
I can confirm that 1.9.6 does, indeed, address this specific problem.

However, I also see some other problems related to interpolated strings, which I will report in a new ticket.

Thank you,

Dennis Jones
